Index Writer/Reader

jigold opened the pull request:
Still to do:
- What should the index file look like for 0 keys? I think it should be 0 bytes and the reader needs to look at the nKeys in the metadata first.
- Incorporate into LoadBGEN

cseed commented on src/main/scala/is/hail/io/index/IndexWriter.scala:
```diff
@@ -0,0 +1,165 @@
+      val node = internalNodes(level)
+      if (node.size > 0) {
+        val (fileOffset, firstIndex, firstKey, firstKeyOffset) = writeInternalNode(node)
+        flushInternalNodes(level + 1, fileOffset, firstIndex, firstKey, firstKeyOffset)
```
Hmm, I'm confused by this.  You have two loops: the loop here over level and the recursive loop in flushInternalNodes that also seems to be over level.  I feel like you only need one of these.

In particular, to flush, don't you need to write out the leaf, add that node to the first internal node, write it out, add that to the next internal node, etc.?

jigold replied:
I changed the code to hopefully make it clearer that there aren't actually two loops through the levels. The behavior of the code hasn't changed.

cseed commented on src/main/scala/is/hail/io/index/IndexWriter.scala:
```diff
@@ -0,0 +1,165 @@
+    var level = 0
+    while (level < nInternalLevels) {
+      val node = internalNodes(level)
+      if (node.size > 0) {
```
Can this be an assert?  Shouldn't the internal nodes always be non-empty?

jigold replied:
No, not necessarily. You can have internal nodes close to the leaf node that are empty, but higher level ones still need to be flushed. That's why I look for the first internal node that needs to be flushed.

cseed commented on src/main/scala/is/hail/io/index/IndexWriter.scala:
```diff
@@ -0,0 +1,165 @@
+    region.clear()
+    node.clear()
+
+    (fileOffset, firstIndex, firstKey, firstKeyOffset)
```
I feel you pack and unpack these in a few places.  Would be simpler to make a case class for this?  Or just reuse Internal/LeafChild?

jigold commented on src/main/scala/is/hail/io/index/IndexWriter.scala:
```diff
@@ -0,0 +1,165 @@
+  path: String,
+  keyType: Type,
+  branchingFactor: Int = 1024,
+  attributes: Map[String, Any] = Map.empty[String, Any]) {
```
For BGEN loading, we can check the reference genome, contig recoding, and skip invalid loci are set to the same parameters we had when creating the index file.

cseed replied:
Ah, right!  Thanks for clarifying.

jigold commented:
I still have the outstanding question about what to do if the nKeys == 0.

jigold commented on src/main/scala/is/hail/io/index/IndexReader.scala:
```diff
@@ -0,0 +1,157 @@
+    LeafNode(SafeRow(leafType, rv))
+  }
+
+  private def queryByKey(key: Annotation, depth: Int, nodeOffset: Long, ordering: ExtendedOrdering): Option[Long] = {
```
I have a second branch with queryByKey (removed it in this branch) because the code was more complicated and I felt it was better to be looked at separately.

jigold commented:
@cseed Addressed your comments except I pulled queryByKey into a different branch. Two things:

1. What should `path` look like in the metadata from IndexWriter? Right now, I have `path + "/index"`. It's not used anywhere. I hardcoded the paths in IndexReader.

2. I don't use the `firstKeyOffset` in the internal nodes anywhere. What were you envisioning it would be used for? Otherwise, I think we should delete it.

cseed commented:
> Would it just be index which is the correct path relative to the metadata file?

Yes.  So you should take the index path in the metadata interpreted with respect to the base of the index file (directory) rather than hardcoding it.  This might seem like overkill but this flexibility has proved useful for the Matrix/Table format and I'd like to copy the idiom here.

 > I don't use the firstKeyOffset in the internal nodes anywhere. What were you envisioning it would be used for? Otherwise, I think we should delete it.

So having indices like this effectively make a Matrix/Table arbitrarily repartitionable.  For example, we can double the partitioning for free by splitting one partition into two by splitting at the roughly the midpoint row which I want to get from the root block.  That's what I was thinking firstKeyOffset would be used for.  This would be for a rough split.  For a split accurate to the row-level, we'd have to read down to the leaf node blocks.  Maybe this isn't really worth it.  I'm on the fence.  Delete it if you want.

I thought of one more thing we should support in the file: store arbitrary user annotations of the keys.  If the annotation is `+struct {}` it would have no overhead.  This will give us some future flexibility and I can imagine the following use case: When we go to a sparse VCF, we'll want to store "checkpoint" rows to avoid having to search backward arbitrarily far to find ref blocks that might overlap with our variant of interest.  The alternative is to make the first row of each partition a checkpoint block, but in that case, we can no longer seek into the middle of the partition if we want to track overlapping ref blocks.  So we want to search for the first checkpoint row before our row of interest.  Can we add this to the format but not use it?

cseed commented on src/main/scala/is/hail/io/index/IndexWriter.scala:
```diff
@@ -0,0 +1,176 @@
+      }
+    }
+
+    val idx = elementIdx - leafNode.size
```
I think when you create/clear a leaf node builder, it should know its first key index.  Then you don't need to calculate it here or pass it to LeafNodeBuilder.write.

cseed commented on src/main/scala/is/hail/io/index/IndexWriter.scala:
```diff
@@ -0,0 +1,176 @@
+      val node = internalNodes(level)
+
+      if (node.size == 0) {
+        node.setFirstIndex(info.firstIndex)
```
As for LeafNodeBuilder, I'd put the index in the InternalNodeBuilder constructor/clear.

cseed commented on src/main/scala/is/hail/io/index/IndexWriter.scala:
```diff
@@ -0,0 +1,176 @@
+    val nInternalLevels = math.max(calcDepth - 1, 1) // ensure always one internal node
+
+    def flushInternalNodes(level: Int, info: IndexNodeInfo) {
+      if (level != nInternalLevels) {
```
I think there is a potential bug here: flushing could cause the tree to grow one level.

I think I'd integrate this code with write: have write take a flush: Boolean argument, and keep going up the tree if the block is full or flush is true and have write skip over empty blocks.  Finally, have write return the offset of the last block it writes, which you can use as the rootOffset in the case flush = true.

jigold commented on src/main/scala/is/hail/io/index/IndexWriter.scala:
```diff
@@ -0,0 +1,176 @@
+
+  private def writeMetadata() = {
+    hConf.writeTextFile(metadataFile) { out =>
+      val metadata = IndexMetadata(1, branchingFactor, height, keyType._toPretty, elementIdx, indexFile, rootOffset, attributes)
```
I can't find `toParsableString` anywhere.

jigold commented on src/main/scala/is/hail/io/index/IndexWriter.scala:
```diff
@@ -0,0 +1,176 @@
+  private def calcDepth: Int =
+    math.max(1, (math.log10(elementIdx) / math.log10(branchingFactor)).ceil.toInt) // max necessary for array of length 1 becomes depth = 0
+
+  private def height: Int = math.max(calcDepth - 1, 1) + 1 // ensure always at least one internal level and one leaf level
```
The last internal node is one that is setup but never written. Otherwise, you don't have access to the first child by the time you're ready to add an extra level. So the rule is actually `height = internalNodes.length`.

cseed commented on src/main/scala/is/hail/io/index/IndexWriter.scala:
```diff
@@ -0,0 +1,164 @@
+
+    if (leafNodeBuilder.size == branchingFactor || (flush && leafNodeBuilder.size > 0)) {
+      val info = writeLeafNode()
+      if (internalNodeBuilders.isEmpty)
```
The root block has to be internal.  I think you should always have an internalLeafNode.  That will  elegantly handle the case of an empty index.

cseed commented on src/main/scala/is/hail/io/index/IndexWriter.scala:
```diff
@@ -0,0 +1,164 @@
+        internalNodeBuilders += new InternalNodeBuilder(keyType, annotationType, info.firstIndex)
+      internalNodeBuilders(0) += info
+      writeInternalNodes(0)
+    } else if (flush && internalNodeBuilders.size > 0)
```
If you require an internal block, you can eliminate the second half of this condition.

jigold commented on src/main/scala/is/hail/io/index/IndexReader.scala:
```diff
@@ -0,0 +1,138 @@
+  var cacheHits = 0L
+  var cacheMisses = 0L
+
+  @transient private[this] lazy val cache = new util.LinkedHashMap[Long, IndexNode](cacheCapacity, 0.75f, true) {
```
I copied the code from the FASTA Reader. I looked at the original PR and @danking's comment was

>>Is there a compelling reason not to use lazy val here? That's basically scala's built-in, thread-safe way to lazily initialize a value.
